APMrover2/mode_breadcrumb.cpp

```diff
@@ -0,0 +1,183 @@
+#include "mode.h"
+#include "Rover.h"
+
+
+bool ModeBreadcrumb::_enter()
+{
+    //don't call enter again if _crumbs exists
+    if (_crumbs != nullptr) {
+        return false;
+    }
+
+    if (_crumbs_sem == nullptr){
+        // create semaphore
+        _crumbs_sem = hal.util->new_semaphore();
+        if (_crumbs_sem == nullptr) {
+            return false;
+        }
+    }
+
+    _current_crumb = nullptr;
+
+    // allocate crumbs queue
+    _crumbs = (Location*)calloc(crumbs_max, sizeof(Location));
+
+    _crumbs_tail = 0;
+    _crumbs_head = 0;
+
+    lonely_mode = nullptr;
+
+    //reuse guided mode
+    if (!ModeGuided::_enter()) {
+        free(_crumbs);
+        return false;
```

Comment:
_crumbs is not being set to `nullptr` here.

Definitely time to think about a `goto` to handle cleanup :-)